apiserver/facades/client/charms/client.go

```diff
@@ -531,3 +550,147 @@ func (a *API) IsMetered(args params.CharmURL) (params.IsMeteredResult, error) {
 	}
 	return params.IsMeteredResult{Metered: false}, nil
 }
+
+// CheckCharmPlacement isn't on the v13 API.
+func (a *APIv3) CheckCharmPlacement(_, _ struct{}) {}
+
+// CheckCharmPlacement checks if a charm is allowed to be placed with in a
+// given application.
+func (a *API) CheckCharmPlacement(args params.ApplicationCharmPlacements) (params.ErrorResults, error) {
+	if err := a.checkCanRead(); err != nil {
+		return params.ErrorResults{}, errors.Trace(err)
+	}
+
+	results := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Placements)),
+	}
+	for i, placement := range args.Placements {
+		result, err := a.checkCharmPlacement(placement)
+		if err != nil {
+			return params.ErrorResults{}, errors.Trace(err)
+		}
+		results.Results[i] = result
+	}
+
+	return results, nil
+}
+
+func (a *API) checkCharmPlacement(arg params.ApplicationCharmPlacement) (params.ErrorResult, error) {
+	curl, err := charm.ParseURL(arg.CharmURL)
+	if err != nil {
+		return params.ErrorResult{
+			Error: apiservererrors.ServerError(err),
+		}, nil
+	}
+
+	// The placement logic below only cares about charmhub charms. Once we have
+	// multiple architecture support for charmhub, we can remove the placement
+	// check.
+	if !charm.CharmHub.Matches(curl.Schema) {
+		return params.ErrorResult{}, nil
+	}
+
+	// Get the application if it's not found, just return without an error as
+	// the charm can be placed in the application once it's created.
+	app, err := a.backendState.Application(arg.Application)
+	if errors.IsNotFound(err) {
+		return params.ErrorResult{}, nil
+	} else if err != nil {
+		return params.ErrorResult{
+			Error: apiservererrors.ServerError(err),
+		}, nil
+	}
+
+	// We don't care for subordinates here.
+	if !app.IsPrincipal() {
+		return params.ErrorResult{}, nil
+	}
+
+	constraints, err := app.Constraints()
```

Comment:
Do you get a blank constraints.Value here if the constraints are not found or a nil value? If it's the latter, if you get ErrNotFound you will skip the next if block and panic on L619

Reply:
constraints.Value is a struct so you would get an empty struct, essentially everything is nil inside, so you have to check HasArch everywhere.

Reply:
Ah cool then. Thought it could be a pointer.